x11rb-async/src/rust_connection/mod.rs

```diff
@@ -0,0 +1,1244 @@
+                }
+
+                // Register a listener for the reply.
+                let listener = self.new_input.listen();
+
+                // Maybe a packet was delivered while we were registering the listener.
+                if let Some(result) = get_result() {
+                    return result;
+                }
+
+                // Wait for the next packet.
+                listener.await;
+            }
+        })
+    }
+
+    fn prefetch_maximum_request_bytes(
+        &self,
+    ) -> Pin<Box<dyn futures_lite::Future<Output = ()> + Send + '_>> {
+        Box::pin(async move {
+            self.prefetch_len_impl()
+                .await
+                .expect("Failed to prefetch maximum request bytes");
+        })
+    }
+
+    fn maximum_request_bytes(
+        &self,
+    ) -> Pin<Box<dyn futures_lite::Future<Output = usize> + Send + '_>> {
+        Box::pin(async move {
+            let setup_len = || {
+                self.setup()
+                    .maximum_request_length
+                    .try_into()
+                    .ok()
+                    .and_then(|x: usize| x.checked_mul(4))
+                    .unwrap_or(std::usize::MAX)
+            };
+
+            let mut mrl = match self
+                .prefetch_len_impl()
+                .await
+                .expect("Failed to prefetch maximum request bytes")
+            {
+                Some(mrl) => mrl,
+                None => {
+                    // We must be writing to it at the moment.
+                    return setup_len();
+                }
+            };
+
+            // Complete the prefetching.
+            match *mrl {
+                MaxRequestBytes::Known(len) => len,
+                MaxRequestBytes::Unknown => unreachable!("We are in the Some branch"),
+                MaxRequestBytes::Requested(cookie) => {
+                    let cookie = match cookie {
+                        Some(cookie) => cookie,
+                        None => {
+                            // Not available.
+                            return setup_len();
+                        }
+                    };
+
+                    // Wait for the reply.
+                    let cookie = Cookie::<'_, _, EnableReply>::new(self, cookie);
+
+                    let reply = cookie.reply().await.expect("Failed to get reply");
+
+                    // Mark the request as done.
+                    let total = reply
+                        .maximum_request_length
+                        .try_into()
+                        .ok()
+                        .and_then(|x: usize| x.checked_mul(4))
+                        .unwrap_or(std::usize::MAX);
+
+                    *mrl = MaxRequestBytes::Known(total);
+                    total
+                }
+            }
+        })
+    }
+
+    fn parse_error(&self, error: &[u8]) -> Result<x11rb::x11_utils::X11Error, ParseError> {
+        let extensions = future::block_on(self.extensions.read());
```

Comment:
Couldn't this block in an async context? Perhaps this needs to be made a blocking mutex that is then not held across await points and thus we get.... some complicated code to recheck things after an await. Yay...

I guess this can be left as-is for now.

Reply:
I'll replace this with `read_blocking` once smol-rs/async-lock#39 lands.

Reply:
Thanks!

But, actually... perhaps "this mutex is only read for short amount of times" is enough here. After all, this is "just" parsing one packet and needs to look up few things in the extensions information. I'll mark this conversation as resolved.